internal/engine/dockerprune/docker_pruner.go

```diff
@@ -0,0 +1,176 @@
+package dockerprune
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+
+	"github.com/windmilleng/tilt/internal/docker"
+	"github.com/windmilleng/tilt/internal/sliceutils"
+	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/pkg/logger"
+)
+
+// How often to prune Docker images while Tilt is running
+const dockerPruneInterval = time.Hour
+
+// Prune Docker objects older than this
+// TODO: configurable from Tiltfile for special cases
+const defaultMaxAge = time.Hour * 6
+
+type DockerPruner struct {
+	dCli docker.Client
+
+	started            bool
+	disabledForTesting bool
+}
+
+var _ store.Subscriber = &DockerPruner{}
+
+func NewDockerPruner(dCli docker.Client) *DockerPruner {
+	return &DockerPruner{dCli: dCli}
+}
+
+func (dp *DockerPruner) DisableForTesting() {
+	dp.disabledForTesting = true
+}
+
+func (dp *DockerPruner) OnChange(ctx context.Context, st store.RStore) {
+	if dp.disabledForTesting {
+		return
+	}
+
+	if dp.started {
+		return
+	}
+
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	if state.DockerPruneSettings.Disabled {
+		return
+	}
+
+	if err := dp.sufficientVersionError(); err != nil {
+		return
+	}
+
+	// wait until state has been kinda initialized, and there's at least one Docker build
+	if !state.TiltStartTime.IsZero() && state.LastTiltfileError() == nil && state.HasDockerBuild() {
+		dp.started = true
+		go func() {
+			select {
+			case <-time.After(time.Minute * 2):
+				dp.Prune(ctx, defaultMaxAge) // report once pretty soon after startup...
+			case <-ctx.Done():
+				return
+			}
+
+			for {
+				select {
+				case <-time.After(dockerPruneInterval):
+					// and once every <interval> thereafter
+					dp.Prune(ctx, defaultMaxAge)
+				case <-ctx.Done():
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (dp *DockerPruner) Prune(ctx context.Context, maxAge time.Duration) {
+	// For future: dispatch event with output/errors to be recorded
+	//   in engineState.TiltSystemState on store (analogous to TiltfileState)
+	err := dp.prune(ctx, maxAge)
+	if err != nil {
+		logger.Get(ctx).Infof("[Docker Prune] error running docker prune: %v", err)
+	}
+}
+
+func (dp *DockerPruner) prune(ctx context.Context, maxAge time.Duration) error {
+	l := logger.Get(ctx)
+	if err := dp.sufficientVersionError(); err != nil {
+		l.Debugf("[Docker Prune] skipping Docker prune, Docker API version too low:\t%v", err)
+		return nil
+	}
+
+	f := filters.NewArgs(
+		filters.Arg("label", docker.BuiltByTiltLabelStr),
+		filters.Arg("until", maxAge.String()),
+	)
+
+	opts := types.BuildCachePruneOptions{
+		All:         true,
+		KeepStorage: 524288000, // 500MB -- TODO: make configurable
```

Comment:
1. Is this required? Or can we leave it unset so that we don't muck with it? It seems like setting this is getting out of scope of pruning and maybe breaking something the user has configured elsewhere. Though, I can't find documentation describing what this does, so....I don't know if this is, like, saying "it's ok if you delete all but 500mb".
2. It's generally clearer / less bug-prone to write numbers like this as 500 * 1024 * 1024

Reply:
fair nough